design/0001-frr.md

```diff
@@ -0,0 +1,411 @@
+# MLB-0001: BGP FRR enablement
+
+## Summary
+
+The purpose of this enhancement is to use [Free Range
+Routing](https://frrouting.org/) (FRR) as an alternative BGP implementation
+in MetalLB. When directed to, MetalLB will publish prefixes via FRR rather than
+MetalLB’s current built-in BGP speaker.
+
+## Motivation
+
+The motivation for this enhancement is to provide an alternative
+production-ready BGP agent for use in MetalLB. Overall, this should reduce the
+maintenance cost for the MetalLB project by reducing the effort required for
+adding additional BGP features and bug fixes. For example, there are a number of
+issues in the current backlog that may be addressed by using FRR:
+
+* [Add support for MP BGP encoding for IPv4 and IPv6](https://github.com/metallb/metallb/pull/590)
+* [BFD support](https://github.com/metallb/metallb/issues/396)
+* [BGP Failover too slow](https://github.com/metallb/metallb/issues/485)
+* [OSPF Support](https://github.com/metallb/metallb/issues/256)
+
+FRR is a mature Linux Foundation routing protocol suite based on Quagga that has
+been used in many production deployments. As such, it has been proven in terms
+of its maturity, flexibility (as can be seen by the [broad range of features](http://docs.frrouting.org/en/latest/bgp.html)
+it supports), scalability, security, reliability and performance. It also
+provides detailed logging features to aid debugging.
+
+The current native BGP implementation of MetalLB has limitations when coexisting
+with other BGP agents on the node. If these BGP agents are attempting to peer
+with external routers, this may invoke BGP’s conflict resolution algorithm. This
+is detailed [here](https://metallb.universe.tf/configuration/calico/) for the
+Calico case and [here](https://metallb.universe.tf/configuration/kube-router/)
+for the Kube-Router case. This enhancement may be a step towards resolving these
+issues by potentially allowing MetalLB to reuse a BGP agent already in use on
+the node. This would enable the local BGP agent (FRR) to resolve any conflicts
+locally before advertising to a remote BGP peer and, therefore, becoming the
+single point-of-contact with that peer.
+
+### Goals
+
+* Provide a configuration option to allow use of FRR as a BGP agent in a MetalLB
+  deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation.
```

Comment:
What is our long-term vision in terms of feature parity between the two BGP implementations? Do we aim to retire the native BGP stack (e.g. after we gain confidence with FRR)?
I can already see a fairly immediate scenario where we want to expose a new knob in the config which turns on a feature which exists in FRR but not in the native BGP stack. I'd say we need a strategy for handling that...

Reply:
I will address this by adding a paragraph describing the roadmap towards this retirement.